src/coprocessor/dag/batch/executors/util/aggr_executor.rs

```diff
@@ -0,0 +1,316 @@
+    ///
+    /// Implementors should update the aggregate function states according to the data of
+    /// these columns.
+    fn process_batch_input(
+        &mut self,
+        entities: &mut Entities<Src>,
+        input: LazyBatchColumnVec,
+    ) -> Result<()>;
+
+    /// Iterates aggregate function states for each group.
+    ///
+    /// Implementors should call `iteratee` for each group with the aggregate function states of
+    /// that group as the argument.
+    ///
+    /// Implementors may return the content of each group as extra columns in the return value
+    /// if there are group by columns.
+    fn iterate_each_group_for_aggregation(
+        &mut self,
+        entities: &mut Entities<Src>,
+        iteratee: impl FnMut(&mut Entities<Src>, &[Box<dyn AggrFunctionState>]) -> Result<()>,
+    ) -> Result<Vec<LazyBatchColumn>>;
+}
+
+/// Some common data that need to be accessed by both `AggregationExecutor`
+/// and `AggregationExecutorImpl`.
+pub struct Entities<Src: BatchExecutor> {
+    pub src: Src,
+    pub context: EvalContext,
+
+    /// The schema of the aggregation executor. It consists of aggregate result columns and
+    /// group by columns.
+    pub schema: Vec<FieldType>,
+
+    /// The aggregate function.
+    pub each_aggr_fn: Vec<Box<dyn AggrFunction>>,
+
+    /// The (output result) cardinality of each aggregate function.
+    pub each_aggr_cardinality: Vec<usize>,
+
+    /// The (input) expression of each aggregate function.
+    pub each_aggr_exprs: Vec<RpnExpression>,
+
+    /// The eval type of the result columns of all aggregate functions. One aggregate function
+    /// may have multiple result columns.
+    pub all_result_column_types: Vec<EvalType>,
+}
+
+/// A shared executor implementation for simple aggregation, hash aggregation and
+/// stream aggregation. Implementation differences are further given via `AggregationExecutorImpl`.
+pub struct AggregationExecutor<
+    C: ExecSummaryCollector,
+    Src: BatchExecutor,
+    I: AggregationExecutorImpl<Src>,
+> {
+    summary_collector: C,
+    imp: I,
+    is_ended: bool,
+    entities: Entities<Src>,
+}
+
+impl<C: ExecSummaryCollector, Src: BatchExecutor, I: AggregationExecutorImpl<Src>>
+    AggregationExecutor<C, Src, I>
+{
+    pub fn new(
+        summary_collector: C,
+        mut imp: I,
+        src: Src,
+        config: Arc<EvalConfig>,
+        aggr_defs: Vec<Expr>,
+        aggr_def_parser: impl AggrDefinitionParser,
+    ) -> Result<Self> {
+        let aggr_fn_len = aggr_defs.len();
+        assert!(aggr_fn_len > 0);
+
+        let src_schema = src.schema();
+        let src_schema_len = src_schema.len();
+
+        let mut schema = Vec::with_capacity(aggr_fn_len * 2);
+        let mut each_aggr_fn = Vec::with_capacity(aggr_fn_len);
+        let mut each_aggr_cardinality = Vec::with_capacity(aggr_fn_len);
+        let mut each_aggr_exprs = Vec::with_capacity(aggr_fn_len * 2);
```

Comment:
If we only support one parameter agg functions now, why we need a capacity of `aggr_fn_len * 2`?

Reply:
nice catch. It's a mistake, should be only `aggr_fn_len`.